src/execution.rs

```diff
@@ -114,6 +123,97 @@ pub trait Report {
     fn start(&mut self, task_name: String);
     fn report(&mut self, result: &CompletedTask);
     fn done(&mut self);
+
+    fn stage(&mut self, full_name: &Vec<String>, stage_rep: StageReport) {
+        let mut full_name = full_name.clone();
+        match stage_rep {
+            StageReport::Start { stage_name } => {
+                full_name.push(stage_name);
+                self.start(full_name.join("::"));
+            }
+            StageReport::Stop {
+                stage_name,
+                status,
+                duration,
+            } => {
+                full_name.push(stage_name);
+                let completed_task = CompletedTask {
+                    full_name,
+                    duration,
+                    stdout: Vec::new(),
+                    stderr: Vec::new(),
+                    status: Status::from(status),
+                };
+                self.report(&completed_task);
+            }
+        }
+    }
+}
+
+pub struct StageReportSender {
+    sender: pipe::Sender,
+    // Keeping the stages that have been "started" around makes it easy
+    // to add a duration later on.
+    started_stages: BTreeMap<String, SystemTime>,
+}
+
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub enum StageStatus {
+    Success,
+    Failure(i32),
+    Skipped(String),
+}
+
+impl From<StageStatus> for Status {
+    fn from(s: StageStatus) -> Self {
+        match s {
+            StageStatus::Success => Status::Success,
+            StageStatus::Failure(code) => Status::Failure(code),
+            StageStatus::Skipped(reason) => Status::Skipped(reason),
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+pub enum StageReport {
+    Start {
+        stage_name: String,
+    },
+    Stop {
+        stage_name: String,
+        status: StageStatus,
+        duration: Duration,
+    },
+}
```

Comment:
Not sure having 2 stage events is a good choice, this opens a lot of questions.

Do we allow nesting? E.g.
```
Start X
  > Start Y
  > Stop Y
  > Start Z
  > Stop Z
Stop X
```

You can have all sorts of weird cases (stop before start, improper nesting, unclosed stages etc.)

I'd prefer if we had just a single event: "stage completed".
This way stages are always linear and easy to reason about.

Reply:
That makes sense. Implemented.